tests/components/dsmr/test_sensor.py

```diff
@@ -349,6 +350,28 @@ async def test_connection_errors_retry(hass, monkeypatch, mock_connection_factor
     assert first_fail_connection_factory.call_count >= 2, "connecting not retried"
 
 
+async def test_connection_abort_error_retry(hass, monkeypatch, mock_connection_factory):
+    """Connection should be retried on connection abort error during setup."""
+    (connection_factory, transport, protocol) = mock_connection_factory
+
+    config = {"platform": "dsmr", "reconnect_interval": 0}
+
+    # override the mock to have it fail the first time and succeed after
+    first_fail_connection_factory = tests.async_mock.AsyncMock(
+        return_value=(transport, protocol),
+        side_effect=chain([OSError(errno.ECONNABORTED)], repeat(DEFAULT)),
+    )
+
+    monkeypatch.setattr(
+        "homeassistant.components.dsmr.sensor.create_dsmr_reader",
+        first_fail_connection_factory,
+    )
+    await async_setup_component(hass, "sensor", {"sensor": config})
+
+    # wait for sleep to resolve
+    await hass.async_block_till_done()
+
```

Comment:
You have no assertions, doesn't this test always pass?

Reply:
The test does fail correctly with the old implementation (by luck)
```
ERROR    homeassistant:runner.py:115 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/dsmr/sensor.py", line 143, in connect_and_reconnect
    transport, protocol = await hass.loop.create_task(reader_factory())
  File "/usr/local/lib/python3.8/unittest/mock.py", line 2134, in _execute_mock_call
    raise result
OSError: 103
```
But you are right, the last line dropped of for some reason. I'll update accordingly